http/server.go

```diff
@@ -26,44 +30,122 @@ const (
 	catchupExpiryFactor = 2
 	roundNumBase        = 10
 	roundNumSize        = 64
+	chainHashParamKey   = "chainHash"
+	roundParamKey       = "round"
 )
 
 var (
 	// Timeout for how long to wait for the drand.PublicClient before timing out
 	reqTimeout = 5 * time.Second
 )
 
+// DrandHandler keeps the reference to the real http handler used by the server
+// to attend to new request, as well as the slice of logic handlers used to process
+// a request. Each handler will attend to requests for one beacon process. The chain hash
+// is used as key.
+type DrandHandler struct {
+	httpHandler http.Handler
+	beacons     map[string]*beaconHandler
+
+	timeout time.Duration
+	context context.Context
+	log     log.Logger
+	version string
+	state   sync.Mutex
+}
+
+type beaconHandler struct {
+	// NOTE: should only be accessed via getChainInfo
+	chainInfo   *chain.Info
+	chainInfoLk sync.RWMutex
+	log         log.Logger
+
+	// Client to handle beacon
+	client client.Client
+
+	// synchronization for blocking writes until randomness available.
+	pendingLk   sync.RWMutex
+	startOnce   sync.Once
+	pending     []chan []byte
+	context     context.Context
+	latestRound uint64
+	version     string
+}
+
 // New creates an HTTP handler for the public Drand API
-func New(ctx context.Context, c client.Client, version string, logger log.Logger) (http.Handler, error) {
+func New(ctx context.Context, c client.Client, version string, logger log.Logger) (*DrandHandler, error) {
 	if logger == nil {
 		logger = log.DefaultLogger()
 	}
-	handler := handler{
-		timeout:     reqTimeout,
-		client:      c,
-		chainInfo:   nil,
-		log:         logger,
-		pending:     nil,
-		context:     ctx,
-		latestRound: 0,
-		version:     version,
+
+	handler := &DrandHandler{
+		timeout: reqTimeout,
+		log:     logger,
+		context: ctx,
+		version: version,
+		beacons: make(map[string]*beaconHandler),
 	}
 
-	mux := http.NewServeMux()
-	//TODO: aggregated bulk round responses.
+	mux := chi.NewMux()
```

Comment:
what are you getting out of chi mux versus the stdlib one?

you're not refactoring the routes to idiomatic chi (`mux.With(commonHeaders).Get('/health', handler.Health)`.
The templating of the `chainHashParamKey` below feels suspect.

Reply:
The main advantage that we make use in `chi` is the param extract util with `chi.URLParam(r, chainHashParamKey)`.
Since it's compatible with stdlib, we saw no need to change the already established routes.
I can change them to be chi idiomatic as you wrote if that is the case.